### ISSUE TITLE
vs2017 projects cannot update PDB with "Symbols" flags

### ISSUE THREAD

**Issue by zao** [bkaradzic/GENie#266]
Tested with GENie from trunk: `GENie - Project generator tool version 738 (commit 4baa591b3c6ecf89904e85169dfc8ab068816250)`

Contents of `genie.lua`:

    solution "vs_repro"
      configurations { "Debug" }
      targetdir "bin"
      language "C++"
      flags { "Symbols" }
    
      project "dummy"
        kind "SharedLib"
        files { "foo.cc" }

Generate build scripts with `genie vs2017`.
Build project in VS 2017 RC, and then recompile any source file results in:

    1>------ Build started: Project: dummy, Configuration: Debug Win32 ------
    1>foo.cc
    1>f:\temp\genie-vs-repro\foo.cc : fatal error C1052: program database file, 'F:\Temp\genie-vs-repro\bin-debug\dummy.pdb', was generated by the linker with /DEBUG:fastlink; compiler cannot update such PDB files; please delete it or use /Fd to specify a different PDB filename
    1>Done building project "dummy.vcxproj" -- FAILED.
    ========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

**Comment by MikePopoloski** [comment]
It looks like this is caused by having the compiler pdb named the same as the linker pdb (two separate output files, one specified with /Fd for the compiler and the other with /PDB for the linker). The compiler PDB isn't used for anything interesting as far as I know (debugging just uses the linker PDB, as it has actual symbols), so as long as it gets a unique name everything should be fine.

Specifying a /Fd under buildoptions in your genie script should do the trick, but unfortunately that produces warnings about /Fd being on the command line twice. Ideally genie should just append something like "_compiler" to the compiler PDB name.

**Comment by beevik** [comment]
Hey @MikePopoloski .. I notice on VS2015 that this change led to solutions that no longer generate `foo.pdb` files with corresponding `foo.lib` files.  Instead we only get `foo_compiler.pdb`.  Unfortunately, this has broken a lot of our scripts that were assuming the existence of a pdb file that had the same name as the lib file.

Can this change be isolated to VS2017?

**Comment by bkaradzic** [comment]
I managed to reproduce this issue with latest VS2017, but then it went away and I can't reproduce it anymore... Anyone actually understand why VS2017 even when /DEBUG:fastlink is not specified does this?

**Comment by bkaradzic** [comment]
It seems only the first project in solution has this issue?!

**Comment by MikePopoloski** [comment]
Oops, sorry about this. I should have tested with VS 2015 before
committing. Maybe we should just gate this change behind a version check? I
was under the understanding that both 2015 and 2017 generated pdbs the same
way, but obviously something else has changed there.

As for why it's only happening sporadically for you, that makes sense since
it's a race condition between the compiler and linker, and even if you
aren't using fastlink I think 2017 tries to do both concurrently.

**Comment by MikePopoloski** [comment]
Note that a full rebuild in 2017 should always cause the issue to occur.
Normal builds might not.

**Comment by bkaradzic** [comment]
I just tested this, and only first project on the top of solution has this issue, everything else is fine. If you create project with single dummy.c file and put it on the top, and then remove .c file from it, everything will work correctly.

**Comment by bkaradzic** [comment]
>  As for why it's only happening sporadically for you

I think I was switching between project, not realizing it. Now I see issue always with top-most project, everything else works fine.

**Comment by bkaradzic** [comment]
Try adding this before any other projects:

```
project "dummy"
    kind "ConsoleApp"
    files { "dummy.c" }
    excludes { "dummy.c" }
```

**Comment by bkaradzic** [comment]
Did a bit more investigation. It looks like if .pdb file name/path stays whatever VS default is, everything works, but if anything is changed it causes this issue.

**Comment by Nuclearfossil** [comment]
Just ran into this issue myself.

I can confirm that manually removing the <ProgramDataBaseFileName> in the project file resolves the issue, but that's not a proper solution.

If you're curious, I have a sample project here (https://github.com/Nuclearfossil/IntroToCPPGraphics) that reproduces the issue.

**Comment by Nuclearfossil** [comment]
So, the issue appears to be (at least, what I've observed) is that setting `<ProgramDataBaseFileName>` in the `<CLCompile>` step stomps on the name that MSBuild is expecting to be able to use in the `<Link>` step. A couple of ideas come to mind:

- when you provide a `Symbols` flag, you have to provide a custom PDB file name that doesn't map to the exe file name
- Modify the Linker parameter to use a different PDB file (ie: `linkoptions {"/PDB:somethingelse.pdb"}`)
- Even if you provide the `Symbols` flag, you don't have to explicitly define the `<ProgramDataBaseFilename>`. Expose a separate option to name either the Compiler and Linker PDB files.

**Comment by amerkoleci** [comment]
Has this issue been fixed? Happens to me too.

**Comment by bkaradzic** [comment]
Nope. Current fix that works for me would be to disallow naming .pdb, but that's not ideal...

**Comment by MikePopoloski** [comment]
So I think my fix works as long as we exclude it for *static library* projects (which is what was breaking things for @beevik ). The fix you just checked in will probably break things for him in the same way when he upgrades to VS 2017. But if it's fixed for now, maybe nobody will worry about it until then.

**Comment by bkaradzic** [comment]
Yeah, I was just jiggling these XML bits until I fixed executable linking error. Maybe complete fix is to toggle those two sections depending of `kind` of project. If it's executable or shared lib it goes to `<Link>` section, if it's library it uses `<ClCompile>` section.

**Comment by tivolo** [comment]
Note that compiler PDB and linker PDB are two completely different things, and should never be the same output file - this leads to increased link times, which has always been the case in all versions before VS 2017, even though it worked.

See https://blogs.msdn.microsoft.com/yash/2007/10/12/pdb-files-what-are-they-and-how-to-generate-them/ and https://public.kitware.com/Bug/view.php?id=14062 for more info.

I would recommend not setting the compiler PDB output (/FD) at all, but rather let VS use its default. The compiler PDB is an intermediate file and not a build artefact.

**Comment by forrestthewoods** [comment]
What's the latest on this issue? Did anyone discover a proper fix?

**Comment by rhoot** [comment]
Pretty sure I fixed this in #358. Anyone able to verify?
